core/blockchain.go

```diff
@@ -3015,3 +3034,155 @@ func EnablePersistDiff(limit uint64) BlockChainOption {
+			hash, err := GetTrustedDiffHash(diff)
+			if err != nil {
+				res.Status = types.StatusUnexpectedError
+				return &res, err
+			}
+
+			diff.DiffHash = hash
+		}
+
+		if diffHash != diff.DiffHash {
+			res.Status = types.StatusDiffHashMismatch
+			return &res, nil
+		}
+
+		header := bc.GetHeaderByHash(blockHash)
+		if header == nil {
+			res.Status = types.StatusUnexpectedError
+			return &res, fmt.Errorf("unexpected error, header not found")
+		}
+		res.Status = types.StatusFullVerified
+		res.Root = header.Root
+		return &res, nil
+	}
+
+	header := bc.GetHeaderByHash(blockHash)
+	if header == nil {
+		if blockNumber > bc.CurrentHeader().Number.Uint64()-11 {
+			res.Status = types.StatusPossibleFork
+			return &res, nil
+		}
+
+		res.Status = types.StatusImpossibleFork
+		return &res, nil
+	}
+
+	res.Status = types.StatusUntrustedVerified
+	res.Root = header.Root
+	return &res, nil
+}
+
+func (bc *BlockChain) GetTrustedDiffLayer(blockHash common.Hash) *types.DiffLayer {
+	var diff *types.DiffLayer
+	if cached, ok := bc.diffLayerCache.Get(blockHash); ok {
+		diff = cached.(*types.DiffLayer)
+		return diff
+	}
+
+	diffStore := bc.db.DiffStore()
+	if diffStore != nil {
+		diff = rawdb.ReadDiffLayer(diffStore, blockHash)
+	}
+	return diff
+}
+
+func (bc *BlockChain) GenerateDiffLayer(blockHash common.Hash) (*types.DiffLayer, error) {
+	if bc.snaps == nil {
+		return nil, fmt.Errorf("snapshot disabled, can't generate difflayer")
+	}
+
+	block := bc.GetBlockByHash(blockHash)
+	if block == nil {
+		return nil, fmt.Errorf("block not found, block number: %d, blockhash: %v", block.NumberU64(), blockHash)
+	}
+
+	parent := bc.GetBlockByHash(block.ParentHash())
+	if parent == nil {
+		return nil, fmt.Errorf("block not found, block number: %d, blockhash: %v", block.NumberU64()-1, block.ParentHash())
+	}
+	statedb, err := bc.StateAt(parent.Root())
+	if err != nil {
+		return nil, fmt.Errorf("state not found for block number (%d): %v", parent.NumberU64(), err)
+	}
+
+	// Replay transactions.
+	signer := types.MakeSigner(bc.Config(), block.Number())
+	for _, tx := range block.Transactions() {
+		msg, _ := tx.AsMessage(signer)
+		txContext := NewEVMTxContext(msg)
+		context := NewEVMBlockContext(block.Header(), bc, nil)
+		vmenv := vm.NewEVM(context, txContext, statedb, bc.Config(), vm.Config{})
+
+		if posa, ok := bc.Engine().(consensus.PoSA); ok {
+			if isSystem, _ := posa.IsSystemTransaction(tx, block.Header()); isSystem {
+				balance := statedb.GetBalance(consensus.SystemAddress)
+				if balance.Cmp(common.Big0) > 0 {
+					statedb.SetBalance(consensus.SystemAddress, big.NewInt(0))
+					statedb.AddBalance(block.Header().Coinbase, balance)
+				}
+			}
+		}
+
+		if _, err := ApplyMessage(vmenv, msg, new(GasPool).AddGas(tx.Gas())); err != nil {
+			return nil, fmt.Errorf("transaction %#x failed: %v", tx.Hash(), err)
+		}
+		statedb.Finalise(vmenv.ChainConfig().IsEIP158(block.Number()))
+	}
+
+	return statedb.GenerateDiffLayer(), nil
```

Comment:
Is this been tested? Since `IntermediateRoot` is required before  `GenerateDiffLayer`.

Reply:
Yes, you are right, I have missed this, I think we just need intermediate account root, state root is not needed.

Reply:
Updated